File: tools/agentstream/checklist/tracker_receipt.go
// Tracker completion receipt parsing for agentstream checklist turns.
// [IMPL-TIED_CHECKLIST_GATE_ENFORCEMENT] [ARCH-TIED_CHECKLIST_GATE_ENFORCEMENT] [REQ-TIED_CHECKLIST_GATE_ENFORCEMENT]
// How: PARSE_TRACKER_COMPLETION_RECEIPT scans fenced JSON for strict agentstream_tracker envelope.
package checklist

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"strings"
)

const trackerReceiptSchemaVersion = 1

// CompletionReceipt is the strict agentstream_tracker completion envelope.
type CompletionReceipt struct {
	SchemaVersion int      `json:"schema_version"`
	Slug          string   `json:"slug"`
	Disposition   string   `json:"disposition"`
	EvidenceRefs  []string `json:"evidence_refs,omitempty"`
	Policy        string   `json:"policy,omitempty"`
	Rationale     string   `json:"rationale,omitempty"`
	Owner         string   `json:"owner,omitempty"`
	Expiry        string   `json:"expiry,omitempty"`
	Approval      string   `json:"approval,omitempty"`
	ResidualRisk  string   `json:"residual_risk,omitempty"`
}

type trackerEnvelope struct {
	Receipt CompletionReceipt `json:"agentstream_tracker"`
}

var allowedReceiptFields = map[string]struct{}{
	"schema_version": {},
	"slug":           {},
	"disposition":    {},
	"evidence_refs":  {},
	"policy":         {},
	"rationale":      {},
	"owner":          {},
	"expiry":         {},
	"approval":       {},
	"residual_risk":  {},
}

var allowedDispositions = map[string]struct{}{
	"completed":      {},
	"not_applicable": {},
	"waived":         {},
}

// ParseTrackerCompletionReceipt scans transcript for the latest valid agentstream_tracker block.
func ParseTrackerCompletionReceipt(transcript, expectedSlug string) (CompletionReceipt, bool, error) {
	blocks := fencedJSONBlocks(transcript)
	for i := len(blocks) - 1; i >= 0; i-- {
		var raw map[string]json.RawMessage
		if err := json.Unmarshal([]byte(blocks[i]), &raw); err != nil {
			if strings.Contains(blocks[i], "agentstream_tracker") {
				return CompletionReceipt{}, false, fmt.Errorf("malformed_receipt: %w", err)
			}
			continue
		}
		body, ok := raw["agentstream_tracker"]
		if !ok {
			continue
		}
		if err := rejectUnknownReceiptFields(body); err != nil {
			return CompletionReceipt{}, false, err
		}
		var env trackerEnvelope
		if err := json.Unmarshal(body, &env.Receipt); err != nil {
			return CompletionReceipt{}, false, fmt.Errorf("malformed_receipt: %w", err)
		}
		if err := ValidateCompletionReceipt(env.Receipt, expectedSlug); err != nil {
			return CompletionReceipt{}, false, err
		}
		return env.Receipt, true, nil
	}
	return CompletionReceipt{}, false, fmt.Errorf("missing_receipt")
}

func rejectUnknownReceiptFields(body json.RawMessage) error {
	var obj map[string]json.RawMessage
	if err := json.Unmarshal(body, &obj); err != nil {
		return fmt.Errorf("malformed_receipt: %w", err)
	}
	for key := range obj {
		if _, ok := allowedReceiptFields[key]; !ok {
			return fmt.Errorf("unknown_field: %q", key)
		}
	}
	return nil
}

// ValidateCompletionReceipt checks schema, slug, and disposition evidence contracts.
func ValidateCompletionReceipt(receipt CompletionReceipt, expectedSlug string) error {
	if receipt.SchemaVersion != trackerReceiptSchemaVersion {
		return fmt.Errorf("unsupported_schema: schema_version must be %d", trackerReceiptSchemaVersion)
	}
	slug := strings.TrimSpace(receipt.Slug)
	if slug == "" {
		return fmt.Errorf("malformed_receipt: slug required")
	}
	if slug != strings.TrimSpace(expectedSlug) {
		return fmt.Errorf("wrong_slug: receipt slug %q does not match expected %q", slug, expectedSlug)
	}
	disp := strings.TrimSpace(receipt.Disposition)
	if disp == "skipped" {
		return fmt.Errorf("skipped_disposition_rejected")
	}
	if _, ok := allowedDispositions[disp]; !ok {
		return fmt.Errorf("invalid_disposition: %q", disp)
	}
	switch disp {
	case "completed":
		if !nonEmptyStringList(receipt.EvidenceRefs) {
			return fmt.Errorf("missing_disposition_evidence: completed requires evidence_refs")
		}
	case "not_applicable":
		if strings.TrimSpace(receipt.Policy) == "" || strings.TrimSpace(receipt.Rationale) == "" {
			return fmt.Errorf("missing_disposition_evidence: not_applicable requires policy and rationale")
		}
	case "waived":
		for _, field := range []struct {
			name string
			val  string
		}{
			{"owner", receipt.Owner},
			{"expiry", receipt.Expiry},
			{"approval", receipt.Approval},
			{"residual_risk", receipt.ResidualRisk},
		} {
			if strings.TrimSpace(field.val) == "" {
				return fmt.Errorf("missing_disposition_evidence: waived requires %s", field.name)
			}
		}
	}
	return nil
}

// ReceiptHash returns a stable hash for idempotent replay detection.
func ReceiptHash(receipt CompletionReceipt) string {
	body, _ := json.Marshal(receipt)
	sum := sha256.Sum256(body)
	return hex.EncodeToString(sum[:])
}

func nonEmptyStringList(items []string) bool {
	for _, item := range items {
		if strings.TrimSpace(item) != "" {
			return true
		}
	}
	return false
}

func fencedJSONBlocks(text string) []string {
	lines := strings.Split(text, "\n")
	var blocks []string
	inFence := false
	var current []string
	for _, line := range lines {
		trimmed := strings.TrimSpace(line)
		if !inFence {
			if trimmed == "```json" {
				inFence = true
				current = current[:0]
			}
			continue
		}
		if strings.HasPrefix(trimmed, "```") {
			blocks = append(blocks, strings.Join(current, "\n"))
			inFence = false
			current = nil
			continue
		}
		current = append(current, line)
	}
	return blocks
}

File: mcp-server/packages/agentstream/src/tracker-receipt.ts
/**
 * [IMPL-TIED_CHECKLIST_GATE_ENFORCEMENT] [ARCH-TIED_CHECKLIST_GATE_ENFORCEMENT] [REQ-TIED_CHECKLIST_GATE_ENFORCEMENT]
 * Tracker completion receipt parse/validate (Go checklist/tracker_receipt.go parity).
 */
import { createHash } from "node:crypto";

export const TRACKER_RECEIPT_SCHEMA_VERSION = 1;

export type TrackerEvidenceRef = string | Record<string, unknown>;

export type TrackerCompletionReceipt = {
  schema_version: number;
  slug: string;
  disposition: string;
  evidence_refs?: TrackerEvidenceRef[];
  policy?: string;
  rationale?: string;
  owner?: string;
  expiry?: string;
  approval?: string;
  residual_risk?: string;
  instruction_nonce?: string;
  instruction_hash?: string;
  request_token?: string;
  run_id?: string;
};

export type IssuedInstruction = {
  nonce: string;
  hash: string;
  requestToken: string;
  runID: string;
};

const allowedReceiptFields = new Set([
  "schema_version",
  "slug",
  "disposition",
  "evidence_refs",
  "policy",
  "rationale",
  "owner",
  "expiry",
  "approval",
  "residual_risk",
  "instruction_nonce",
  "instruction_hash",
  "request_token",
  "run_id",
]);

const allowedDispositions = new Set(["completed", "not_applicable", "waived"]);

function fencedJsonBlocks(text: string): string[] {
  const lines = text.split("\n");
  const blocks: string[] = [];
  let inFence = false;
  let current: string[] = [];
  for (const line of lines) {
    const trimmed = line.trim();
    if (!inFence) {
      if (trimmed === "```json") {
        inFence = true;
        current = [];
      }
      continue;
    }
    if (trimmed.startsWith("```")) {
      blocks.push(current.join("\n"));
      inFence = false;
      current = [];
      continue;
    }
    current.push(line);
  }
  return blocks;
}

function nonEmptyEvidenceRefs(refs: TrackerEvidenceRef[] | undefined): boolean {
  if (!refs || refs.length === 0) {
    return false;
  }
  for (const ref of refs) {
    if (typeof ref === "string" && ref.trim() !== "") {
      return true;
    }
    if (typeof ref === "object" && ref !== null && Object.keys(ref).length > 0) {
      return true;
    }
  }
  return false;
}

function rejectUnknownReceiptFields(body: Record<string, unknown>): void {
  for (const key of Object.keys(body)) {
    if (!allowedReceiptFields.has(key)) {
      throw new Error(`unknown_field: ${JSON.stringify(key)}`);
    }
  }
}

export function validateCompletionReceipt(
  receipt: TrackerCompletionReceipt,
  expectedSlug: string,
): void {
  if (receipt.schema_version !== TRACKER_RECEIPT_SCHEMA_VERSION) {
    throw new Error(
      `unsupported_schema: schema_version must be ${TRACKER_RECEIPT_SCHEMA_VERSION}`,
    );
  }
  const slug = String(receipt.slug ?? "").trim();
  if (slug === "") {
    throw new Error("malformed_receipt: slug required");
  }
  if (slug !== String(expectedSlug).trim()) {
    throw new Error(
      `wrong_slug: receipt slug ${JSON.stringify(slug)} does not match expected ${JSON.stringify(expectedSlug)}`,
    );
  }
  const disp = String(receipt.disposition ?? "").trim();
  if (disp === "skipped") {
    throw new Error("skipped_disposition_rejected");
  }
  if (!allowedDispositions.has(disp)) {
    throw new Error(`invalid_disposition: ${JSON.stringify(disp)}`);
  }
  switch (disp) {
    case "completed":
      if (!nonEmptyEvidenceRefs(receipt.evidence_refs)) {
        throw new Error(
          "missing_disposition_evidence: completed requires evidence_refs",
        );
      }
      break;
    case "not_applicable":
      if (
        String(receipt.policy ?? "").trim() === "" ||
        String(receipt.rationale ?? "").trim() === ""
      ) {
        throw new Error(
          "missing_disposition_evidence: not_applicable requires policy and rationale",
        );
      }
      break;
    case "waived":
      for (const field of [
        "owner",
        "expiry",
        "approval",
        "residual_risk",
      ] as const) {
        if (String(receipt[field] ?? "").trim() === "") {
          throw new Error(
            `missing_disposition_evidence: waived requires ${field}`,
          );
        }
      }
      break;
    default:
      break;
  }
}

export function parseTrackerCompletionReceipt(
  transcript: string,
  expectedSlug: string,
): { receipt: TrackerCompletionReceipt; ok: true } | { ok: false; error: Error } {
  const blocks = fencedJsonBlocks(transcript);
  for (let i = blocks.length - 1; i >= 0; i--) {
    const block = blocks[i]!;
    let raw: Record<string, unknown>;
    try {
      raw = JSON.parse(block) as Record<string, unknown>;
    } catch (err) {
      if (block.includes("agentstream_tracker")) {
        return { ok: false, error: new Error(`malformed_receipt: ${String(err)}`) };
      }
      continue;
    }
    if (!Object.prototype.hasOwnProperty.call(raw, "agentstream_tracker")) {
      continue;
    }
    const body = raw.agentstream_tracker;
    if (body === null || typeof body !== "object" || Array.isArray(body)) {
      return { ok: false, error: new Error("malformed_receipt") };
    }
    try {
      rejectUnknownReceiptFields(body as Record<string, unknown>);
    } catch (err) {
      return { ok: false, error: err instanceof Error ? err : new Error(String(err)) };
    }
    const receipt = body as TrackerCompletionReceipt;
    try {
      validateCompletionReceipt(receipt, expectedSlug);
    } catch (err) {
      return { ok: false, error: err instanceof Error ? err : new Error(String(err)) };
    }
    return { receipt, ok: true };
  }
  return { ok: false, error: new Error("missing_receipt") };
}

export function validateReceiptBinding(
  receipt: TrackerCompletionReceipt,
  issued: IssuedInstruction,
): void {
  if (
    String(receipt.instruction_nonce ?? "").trim() === "" ||
    String(receipt.instruction_hash ?? "").trim() === ""
  ) {
    throw new Error("missing_binding_fields");
  }
  if (
    String(receipt.instruction_nonce ?? "").trim() !== issued.nonce.trim()
  ) {
    throw new Error("stale_instruction_nonce");
  }
  if (String(receipt.instruction_hash ?? "").trim() !== issued.hash.trim()) {
    throw new Error("instruction_hash_mismatch");
  }
  if (
    String(receipt.request_token ?? "").trim() !== issued.requestToken.trim()
  ) {
    throw new Error("request_token_mismatch");
  }
  if (String(receipt.run_id ?? "").trim() !== issued.runID.trim()) {
    throw new Error("run_id_mismatch");
  }
}

export function receiptHash(receipt: TrackerCompletionReceipt): string {
  const body = JSON.stringify(receipt);
  return createHash("sha256").update(body).digest("hex");
}

/** Map receipt to adherence-evidence-resolve CompletionReceipt shape. */
export function receiptForEvidenceResolve(
  receipt: TrackerCompletionReceipt,
): { slug: string; disposition: string; evidenceRefs: (string | Record<string, unknown>)[] } {
  return {
    slug: receipt.slug,
    disposition: receipt.disposition,
    evidenceRefs: receipt.evidence_refs ?? [],
  };
}
